internal/controllers/objectsets/objectset_controller.go

```diff
@@ -0,0 +1,249 @@
+}
+
+type teardownHandler interface {
+	Teardown(
+		ctx context.Context, objectSet genericObjectSet,
+	) (cleanupDone bool, err error)
+}
+
+func NewObjectSetController(
+	c client.Client, log logr.Logger,
+	scheme *runtime.Scheme, dw dynamicCache,
+) *GenericObjectSetController {
+	return newGenericObjectSetController(
+		newGenericObjectSet,
+		newGenericObjectSetPhase,
+		c, log, scheme, dw,
+	)
+}
+
+func NewClusterObjectSetController(
+	c client.Client, log logr.Logger,
+	scheme *runtime.Scheme, dw dynamicCache,
+) *GenericObjectSetController {
+	return newGenericObjectSetController(
+		newGenericClusterObjectSet,
+		newGenericClusterObjectSetPhase,
+		c, log, scheme, dw,
+	)
+}
+
+func newGenericObjectSetController(
+	newObjectSet genericObjectSetFactory,
+	newObjectSetPhase genericObjectSetPhaseFactory,
+	c client.Client, log logr.Logger,
+	scheme *runtime.Scheme, dw dynamicCache,
+) *GenericObjectSetController {
+	controller := &GenericObjectSetController{
+		newObjectSet:      newObjectSet,
+		newObjectSetPhase: newObjectSetPhase,
+
+		client: c,
+		log:    log,
+		scheme: scheme,
+		dw:     dw,
+	}
+
+	phasesReconciler := newPhasesReconciler(c, objectsetphases.NewPhaseReconciler(
+		scheme, c, dw, ownerhandling.NewNative(scheme),
+	), scheme, newObjectSet)
+
+	controller.teardownHandler = phasesReconciler
+
+	controller.reconciler = []reconciler{
+		&revisionReconciler{
+			scheme:       scheme,
+			client:       c,
+			newObjectSet: newObjectSet,
+		},
+		phasesReconciler,
+	}
+
+	return controller
+}
+
+func (c *GenericObjectSetController) SetupWithManager(mgr ctrl.Manager) error {
+	objectSet := c.newObjectSet(c.scheme).ClientObject()
+	objectSetPhase := c.newObjectSetPhase(c.scheme).ClientObject()
+
+	return ctrl.NewControllerManagedBy(mgr).
+		For(objectSet).
+		Owns(objectSetPhase).
+		Watches(c.dw.Source(), &handler.EnqueueRequestForOwner{
+			OwnerType:    objectSet,
+			IsController: false,
+		}).
+		Complete(c)
+}
+
+func (c *GenericObjectSetController) Reconcile(
+	ctx context.Context, req ctrl.Request,
+) (ctrl.Result, error) {
+	log := c.log.WithValues("ObjectSet", req.String())
+	defer log.Info("reconciled")
+	ctx = logr.NewContext(ctx, log)
+
+	objectSet := c.newObjectSet(c.scheme)
+	if err := c.client.Get(
+		ctx, req.NamespacedName, objectSet.ClientObject()); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if meta.IsStatusConditionTrue(*objectSet.GetConditions(), corev1alpha1.ObjectSetArchived) {
+		// We don't want to touch this object anymore.
+		return ctrl.Result{}, nil
+	}
+
+	if !objectSet.ClientObject().GetDeletionTimestamp().IsZero() ||
+		objectSet.IsArchived() {
+		if err := c.handleDeletionAndArchival(ctx, objectSet); err != nil {
+			return ctrl.Result{}, err
+		}
+
+		objectSet.UpdateStatusPhase()
+		// this controller owns status alone, so we can always update it without optimistic locking.
+		objectSet.ClientObject().SetResourceVersion("")
```

Comment:
Why do we set the ResourceVersion to ""?

Reply:
To disable optimistic locking for this patch :)
Otherwise unrelated updates by other clients would return a ResourceVersion conflict error.